synapse/handlers/room_member.py

```diff
@@ -1150,8 +1150,8 @@ async def transfer_room_state_on_room_upgrade(
         logger.info("Transferring room state from %s to %s", old_room_id, room_id)
 
         # Find all local users that were in the old room and copy over each user's state
```

Comment:
We now do what this comment says, but is this comment actually correct?

I don't see anything previously where we checked for `is_mind_id(...)` in the downstream functions here. It feels like we happily copied everyone over.

It makes sense to only copy over the push rules for the local users (`transfer_room_state_on_room_upgrade` -> `copy_user_state_on_room_upgrade` -> `copy_push_rules_from_room_to_room_for_user`)

I'm not sure about `transfer_room_state_on_room_upgrade` -> `copy_user_state_on_room_upgrade` -> `copy_room_tags_and_direct_to_room` though.

Reply:
I'm fairly sure account data (and hence `m.direct` data) and room tags can only be set for local users.
ie. we used to copy over nothing for remote users, so this change looks alright.

Reply:
Woot! Good news to move this forward 🚆

> ie. we used to copy over nothing for remote users, so this change looks alright.

I trust you but can you link where you see this constraint? It's not immediately obvious to me trying to follow some functions down.

Reply:
There's no check in the copying method or its callees. It's just not possible for us to insert account data and room tags for remote users into the database.